File: server/chat_app_server/app/services/v2/message_manager.py
"""
消息管理器
负责处理聊天消息的保存、检索和管理
"""
import time
from typing import Dict, List, Any, Optional

# 导入数据库模型
from ...models.message import MessageCreate


class MessageManager:
    """消息管理器"""
    
    def __init__(self):
        """
        初始化消息管理器
        
        注意：现在直接使用 models 模块中的数据库操作方法
        """
        
        # 缓存最近保存的消息
        self.recent_messages = {}
        
        # 待保存的消息队列
        self.pending_saves = []
        
        # 统计信息
        self.stats = {
            "messages_saved": 0,
            "messages_retrieved": 0,
            "cache_hits": 0,
            "cache_misses": 0
        }
    
    def save_user_message(self, session_id: str, content: str, message_id: str = None) -> Dict[str, Any]:
        """
        保存用户消息
        
        Args:
            session_id: 会话ID
            content: 消息内容
            message_id: 消息ID（可选）
            
        Returns:
            保存的消息信息
        """
        try:
            # 创建消息数据
            message_data = MessageCreate(
                id=message_id,
                sessionId=session_id,
                role="user",
                content=content
            )
            
            # 使用同步方法保存到数据库
            saved_message = MessageCreate.create_sync(message_data)
            
            # 缓存消息
            if saved_message:
                self._cache_message(saved_message)
                self.stats["messages_saved"] += 1
            
            return {
                "success": True,
                "message": saved_message,
                "message_id": saved_message.get("id") if saved_message else None
            }
            
        except Exception as e:
            error_message = f"保存用户消息失败: {str(e)}"
            print(f"Error in save_user_message: {error_message}")
            return {
                "success": False,
                "error": error_message
            }
    
    def save_assistant_message(self, 
                             session_id: str, 
                             content: str, 
                             message_id: str = None,
                             summary: str = None,
                             reasoning: str = None,
                             metadata: Dict[str, Any] = None) -> Dict[str, Any]:
        """
        保存助手消息
        
        Args:
            session_id: 会话ID
            content: 消息内容
            message_id: 消息ID（可选）
            summary: 消息摘要
            reasoning: 推理过程
            metadata: 元数据
            
        Returns:
            保存的消息信息
        """
        try:
            # 创建消息数据
            message_data = MessageCreate(
                id=message_id,
                sessionId=session_id,
                role="assistant",
                content=content,
                summary=summary,
                reasoning=reasoning,
                metadata=metadata
            )
            
            # 使用同步方法保存到数据库
            saved_message = MessageCreate.create_sync(message_data)
            
            # 缓存消息
            if saved_message:
                self._cache_message(saved_message)
                self.stats["messages_saved"] += 1
            
            return {
                "success": True,
                "message": saved_message,
                "message_id": saved_message.get("id") if saved_message else None
            }
            
        except Exception as e:
            error_message = f"保存助手消息失败: {str(e)}"
            print(f"Error in save_assistant_message: {error_message}")
            return {
                "success": False,
                "error": error_message
            }
    
    def save_tool_message(self, 
                         session_id: str, 
                         content: str, 
                         tool_call_id: str,
                         message_id: str = None,
                         metadata: Dict[str, Any] = None) -> Dict[str, Any]:
        """
        保存工具消息
        
        Args:
            session_id: 会话ID
            content: 消息内容
            tool_call_id: 工具调用ID
            message_id: 消息ID（可选）
            metadata: 元数据
            
        Returns:
            保存的消息信息
        """
        try:
            # 创建消息数据
            message_data = MessageCreate(
                id=message_id,
                sessionId=session_id,
                role="tool",
                content=content,
                tool_call_id=tool_call_id,
                metadata=metadata
            )
            
            # 使用同步方法保存到数据库
            saved_message = MessageCreate.create_sync(message_data)
            
            # 缓存消息
            if saved_message:
                self._cache_message(saved_message)
                self.stats["messages_saved"] += 1
            
            return {
                "success": True,
                "message": saved_message,
                "message_id": saved_message.get("id") if saved_message else None
            }
            
        except Exception as e:
            error_message = f"保存工具消息失败: {str(e)}"
            print(f"Error in save_tool_message: {error_message}")
            return {
                "success": False,
                "error": error_message
            }
    
    def get_session_messages(self, session_id: str, limit: int = None) -> List[Dict[str, Any]]:
        """
        获取会话消息
        
        Args:
            session_id: 会话ID
            limit: 消息数量限制
            
        Returns:
            消息列表
        """
        try:
            # 使用同步方法从数据库获取消息
            messages = MessageCreate.get_by_session_sync(session_id, limit)
            
            self.stats["messages_retrieved"] += len(messages)
            
            return messages
            
        except Exception as e:
            error_message = f"获取会话消息失败: {str(e)}"
            print(f"Error in get_session_messages: {error_message}")
            return []
    
    def get_message_by_id(self, message_id: str) -> Optional[Dict[str, Any]]:
        """
        根据ID获取消息
        
        Args:
            message_id: 消息ID
            
        Returns:
            消息信息
        """
        try:
            # 先检查缓存
            if message_id in self.recent_messages:
                self.stats["cache_hits"] += 1
                return self.recent_messages[message_id]
            
            # 从数据库获取
            message = MessageCreate.get_by_id_sync(message_id)
            
            if message:
                self._cache_message(message)
                self.stats["cache_misses"] += 1
                self.stats["messages_retrieved"] += 1
            
            return message
            
        except Exception as e:
            error_message = f"获取消息失败: {str(e)}"
            print(f"Error in get_message_by_id: {error_message}")
            return None
    
    def process_pending_saves(self) -> Dict[str, Any]:
        """
        处理待保存的消息
        
        Returns:
            处理结果
        """
        try:
            if not self.pending_saves:
                return {
                    "success": True,
                    "processed_count": 0,
                    "message": "没有待保存的消息"
                }
            
            processed_count = 0
            errors = []
            
            # 处理所有待保存的消息
            for message_data in self.pending_saves:
                try:
                    saved_message = MessageCreate.create_sync(message_data)
                    if saved_message:
                        self._cache_message(saved_message)
                        processed_count += 1
                        self.stats["messages_saved"] += 1
                except Exception as e:
                    errors.append(f"保存消息失败: {str(e)}")
            
            # 清空待保存队列
            self.pending_saves.clear()
            
            return {
                "success": len(errors) == 0,
                "processed_count": processed_count,
                "errors": errors
            }
            
        except Exception as e:
            error_message = f"处理待保存消息失败: {str(e)}"
            print(f"Error in process_pending_saves: {error_message}")
            return {
                "success": False,
                "error": error_message
            }
    
    def _cache_message(self, message: Dict[str, Any]) -> None:
        """
        缓存消息
        
        Args:
            message: 消息数据
        """
        if message and "id" in message:
            # 限制缓存大小
            if len(self.recent_messages) >= 100:
                # 移除最旧的消息
                oldest_key = next(iter(self.recent_messages))
                del self.recent_messages[oldest_key]
            
            self.recent_messages[message["id"]] = message
    
    def clear_cache(self) -> None:
        """清空缓存"""
        self.recent_messages.clear()
    
    def get_stats(self) -> Dict[str, Any]:
        """
        获取统计信息
        
        Returns:
            统计信息
        """
        return {
            "stats": self.stats.copy(),
            "cache_size": len(self.recent_messages),
            "pending_saves": len(self.pending_saves)
        }

File: server/chat_app_node_server/src/services/v2/message-manager.js
/**
 * 消息管理器 - 负责处理聊天消息的保存、检索和管理
 * 复刻自 Python: app/services/v2/message_manager.py
 */

import { MessageService } from '../../models/message.js';
import { logger } from '../../utils/logger.js';

export class MessageManager {
  constructor() {
    // 缓存最近保存的消息
    this.recentMessages = new Map();

    // 待保存的消息队列
    this.pendingSaves = [];

    // 统计信息
    this.stats = {
      messages_saved: 0,
      messages_retrieved: 0,
      cache_hits: 0,
      cache_misses: 0
    };
  }

  /**
   * 保存用户消息
   */
  saveUserMessage(sessionId, content, messageId = null) {
    try {
      // 创建并保存消息
      const savedMessage = MessageService.createSync({
        id: messageId,
        sessionId,
        role: 'user',
        content
      });

      // 缓存消息
      if (savedMessage) {
        this._cacheMessage(savedMessage);
        this.stats.messages_saved += 1;
      }

      return {
        success: true,
        message: savedMessage,
        message_id: savedMessage?.id
      };
    } catch (error) {
      const errorMessage = `保存用户消息失败: ${error.message}`;
      logger.error(errorMessage, error);
      return {
        success: false,
        error: errorMessage
      };
    }
  }

  /**
   * 保存助手消息
   */
  saveAssistantMessage(sessionId, content, options = {}) {
    try {
      const {
        messageId = null,
        summary = null,
        reasoning = null,
        metadata = null,
        toolCalls = null
      } = options;

      // 创建并保存消息
      const savedMessage = MessageService.createSync({
        id: messageId,
        sessionId,
        role: 'assistant',
        content,
        summary,
        reasoning,
        metadata,
        toolCalls
      });

      // 缓存消息
      if (savedMessage) {
        this._cacheMessage(savedMessage);
        this.stats.messages_saved += 1;
      }

      return {
        success: true,
        message: savedMessage,
        message_id: savedMessage?.id
      };
    } catch (error) {
      const errorMessage = `保存助手消息失败: ${error.message}`;
      logger.error(errorMessage, error);
      return {
        success: false,
        error: errorMessage
      };
    }
  }

  /**
   * 保存工具消息
   */
  saveToolMessage(sessionId, content, toolCallId, options = {}) {
    try {
      const { messageId = null, metadata = null } = options;

      // 创建并保存消息
      const savedMessage = MessageService.createSync({
        id: messageId,
        sessionId,
        role: 'tool',
        content,
        tool_call_id: toolCallId,
        metadata
      });

      // 缓存消息
      if (savedMessage) {
        this._cacheMessage(savedMessage);
        this.stats.messages_saved += 1;
      }

      return {
        success: true,
        message: savedMessage,
        message_id: savedMessage?.id
      };
    } catch (error) {
      const errorMessage = `保存工具消息失败: ${error.message}`;
      logger.error(errorMessage, error);
      return {
        success: false,
        error: errorMessage
      };
    }
  }

  /**
   * 获取会话消息（异步）
   */
  async getSessionMessages(sessionId, limit = null) {
    try {
      const messages = await MessageService.getBySession(sessionId, limit);
      this.stats.messages_retrieved += messages.length;
      return messages;
    } catch (error) {
      logger.error(`获取会话消息失败: ${error.message}`, error);
      return [];
    }
  }

  /**
   * 获取会话消息（同步） - 用于与 Python 版本对齐
   */
  getSessionMessagesSync(sessionId, limit = null) {
    try {
      const db = require('../../models/database-factory.js').getDatabaseSync();
      let query = 'SELECT * FROM messages WHERE session_id = ? ORDER BY created_at ASC';
      const params = [sessionId];

      if (limit) {
        query += ' LIMIT ?';
        params.push(limit);
      }

      const rows = db.fetchallSync(query, params);
      const messages = rows.map(row => {
        return {
          id: row.id,
          sessionId: row.session_id,
          role: row.role,
          content: row.content,
          summary: row.summary,
          toolCalls: row.tool_calls ? JSON.parse(row.tool_calls) : null,
          tool_call_id: row.tool_call_id,
          reasoning: row.reasoning,
          metadata: row.metadata ? JSON.parse(row.metadata) : null,
          created_at: row.created_at
        };
      });

      this.stats.messages_retrieved += messages.length;
      return messages;
    } catch (error) {
      logger.error(`获取会话消息失败: ${error.message}`, error);
      return [];
    }
  }

  /**
   * 根据 ID 获取消息
   */
  async getMessageById(messageId) {
    try {
      // 先检查缓存
      if (this.recentMessages.has(messageId)) {
        this.stats.cache_hits += 1;
        return this.recentMessages.get(messageId);
      }

      // 从数据库获取
      const message = await MessageService.getById(messageId);

      if (message) {
        this._cacheMessage(message);
        this.stats.cache_misses += 1;
        this.stats.messages_retrieved += 1;
      }

      return message;
    } catch (error) {
      logger.error(`获取消息失败: ${error.message}`, error);
      return null;
    }
  }

  /**
   * 处理待保存的消息
   */
  processPendingSaves() {
    try {
      if (this.pendingSaves.length === 0) {
        return {
          success: true,
          processed_count: 0,
          message: '没有待保存的消息'
        };
      }

      let processedCount = 0;
      const errors = [];

      // 处理所有待保存的消息
      for (const messageData of this.pendingSaves) {
        try {
          const savedMessage = MessageService.createSync(messageData);
          if (savedMessage) {
            this._cacheMessage(savedMessage);
            processedCount += 1;
            this.stats.messages_saved += 1;
          }
        } catch (error) {
          errors.push(`保存消息失败: ${error.message}`);
        }
      }

      // 清空待保存队列
      this.pendingSaves = [];

      return {
        success: errors.length === 0,
        processed_count: processedCount,
        errors
      };
    } catch (error) {
      const errorMessage = `处理待保存消息失败: ${error.message}`;
      logger.error(errorMessage, error);
      return {
        success: false,
        error: errorMessage
      };
    }
  }

  /**
   * 缓存消息
   */
  _cacheMessage(message) {
    if (message && message.id) {
      // 限制缓存大小
      if (this.recentMessages.size >= 100) {
        // 移除最旧的消息（Map 的第一个键）
        const firstKey = this.recentMessages.keys().next().value;
        this.recentMessages.delete(firstKey);
      }

      this.recentMessages.set(message.id, message);
    }
  }

  /**
   * 清空缓存
   */
  clearCache() {
    this.recentMessages.clear();
  }

  /**
   * 清空指定会话的缓存消息
   */
  clearCacheForSession(sessionId) {
    try {
      if (this.recentMessages.size === 0) {
        return;
      }

      // 找出属于该会话的消息 ID
      const toDelete = [];
      for (const [messageId, message] of this.recentMessages.entries()) {
        if (message.sessionId === sessionId) {
          toDelete.push(messageId);
        }
      }

      // 删除这些消息
      for (const messageId of toDelete) {
        this.recentMessages.delete(messageId);
      }
    } catch (error) {
      logger.error(`清空会话缓存失败: ${error.message}`, error);
    }
  }

  /**
   * 获取统计信息
   */
  getStats() {
    return {
      stats: { ...this.stats },
      cache_size: this.recentMessages.size,
      pending_saves: this.pendingSaves.length
    };
  }

  /**
   * 获取缓存统计信息（与 AiClient 接口对齐）
   */
  getCacheStats() {
    const sessionCounts = {};

    for (const message of this.recentMessages.values()) {
      const sessionId = message.sessionId;
      if (sessionId) {
        sessionCounts[sessionId] = (sessionCounts[sessionId] || 0) + 1;
      }
    }

    return {
      cache_size: this.recentMessages.size,
      by_session: sessionCounts
    };
  }
}
